src/consumer/consumerGroup.js

```diff
@@ -437,45 +439,48 @@ module.exports = class ConsumerGroup {
         return []
       }
 
-      const results = await Promise.all(requests)
-      return flatten(results)
+      return BufferedAsyncIterator(requests, e => this.recoverFromFetch(e))
     } catch (e) {
-      if (STALE_METADATA_ERRORS.includes(e.type) || e.name === 'KafkaJSTopicMetadataNotLoaded') {
-        this.logger.debug('Stale cluster metadata, refreshing...', {
-          groupId: this.groupId,
-          memberId: this.memberId,
-          error: e.message,
-        })
+      await this.recoverFromFetch(e)
+    }
+  }
 
-        await this.cluster.refreshMetadata()
-        await this.join()
-        await this.sync()
-        throw new KafkaJSError(e.message)
-      }
+  async recoverFromFetch(e) {
+    if (STALE_METADATA_ERRORS.includes(e.type) || e.name === 'KafkaJSTopicMetadataNotLoaded') {
+      this.logger.debug('Stale cluster metadata, refreshing...', {
+        groupId: this.groupId,
+        memberId: this.memberId,
+        error: e.message,
+      })
 
-      if (e.name === 'KafkaJSStaleTopicMetadataAssignment') {
-        this.logger.warn(`${e.message}, resync group`, {
-          groupId: this.groupId,
-          memberId: this.memberId,
-          topic: e.topic,
-          unknownPartitions: e.unknownPartitions,
-        })
+      await this.cluster.refreshMetadata()
+      await this.join()
+      await this.sync()
+      throw new KafkaJSError(e.message)
+    }
 
-        await this.join()
-        await this.sync()
-      }
+    if (e.name === 'KafkaJSStaleTopicMetadataAssignment') {
+      this.logger.warn(`${e.message}, resync group`, {
+        groupId: this.groupId,
+        memberId: this.memberId,
+        topic: e.topic,
+        unknownPartitions: e.unknownPartitions,
+      })
 
-      if (e.name === 'KafkaJSOffsetOutOfRange') {
-        await this.recoverFromOffsetOutOfRange(e)
-      }
+      await this.join()
+      await this.sync()
+    }
 
-      if (e.name === 'KafkaJSBrokerNotFound') {
-        this.logger.debug(`${e.message}, refreshing metadata and retrying...`)
-        await this.cluster.refreshMetadata()
-      }
+    if (e.name === 'KafkaJSOffsetOutOfRange') {
+      await this.recoverFromOffsetOutOfRange(e)
+    }
 
-      throw e
+    if (e.name === 'KafkaJSBrokerNotFound') {
+      this.logger.debug(`${e.message}, refreshing metadata and retrying...`)
+      await this.cluster.refreshMetadata()
     }
+
+    throw e
```

Comment:
For all these cases where we handle the error, shouldn't we return from the function afterwards? Or is there some reason why we still want the error to be re-thrown?

Reply:
We want the error to bubble up and trigger the retry mechanism, the error handling here apply what should happen after the error, the regular flow is responsible for handling the new state.

Reply:
The retry mechanism, of course! Missed that at the late Friday afternoon 😅.